### PR TITLE
Feature: generalised functorialisation

### PR DESCRIPTION
**Pull request by regular-citizen**
This PR implements the recursive machinery for running arbitrary depth functorialisation. It deprecates old functorialisation code and enables functorialisation w.r.t locally maximal variables of arbitrary codimension.

### CODE REVIEW

**Comment by regular-citizen** [comment]
@thibautbenjamin ready for review

**Review by thibautbenjamin — CHANGES_REQUESTED** [review]
Review done. Mostly minor things but a few points that I am not entirely convinced with. Happy to discuss them in person whenever

File: lib/builtin.ml [hunk 1]
```diff
@@ -37,10 +39,44 @@ let comp_n arity =
...
+  if !Settings.explicit_substitutions then
+    (n-1)/2
+  else n
+
 let comp s =
   let arity = arity_comp s in
   comp_n arity
 
+(* returns the n-composite of a (n+j)-cell with a (n+k)-cell *)
+let whisk n j k =
```
[the remark this answers is not in the document]
Comment (regular-citizen) [reply lib/builtin.ml:53]: Not implementing this. Will deal in future PR when shuffling the functorialisation coherence code around

**Comment by regular-citizen** [comment]
Implemented reviews

**Comment by regular-citizen** [comment]
@thibautbenjamin can you merge?

**Review by thibautbenjamin — APPROVED** [review]

**Comment by thibautbenjamin** [comment]
Done! I did squash and merge because the history wasn't very clean

**Pull request merged**